database/skylink.go

```diff
@@ -143,34 +123,130 @@ func (db *DB) AddServerForSkylink(ctx context.Context, skylink string, server st
 		update = bson.M{"$addToSet": bson.M{"servers": server}}
 	}
 	opts := options.Update().SetUpsert(true)
-	_, err = db.staticDB.Collection(collSkylinks).UpdateOne(ctx, filter, update, opts)
+	_, err := db.staticDB.Collection(collSkylinks).UpdateOne(ctx, filter, update, opts)
 	return err
 }
 
 // RemoveServerFromSkylink removes a server to the list of servers known to be
 // pinning this skylink. If the skylink does not exist in the database it will
 // not be inserted.
-func (db *DB) RemoveServerFromSkylink(ctx context.Context, skylink string, server string) error {
-	sl, err := CanonicalSkylink(skylink)
-	if err != nil {
-		return ErrInvalidSkylink
-	}
+func (db *DB) RemoveServerFromSkylink(ctx context.Context, skylink skymodules.Skylink, server string) error {
 	filter := bson.M{
-		"skylink": sl,
+		"skylink": skylink.String(),
 		"servers": server,
 	}
 	update := bson.M{"$pull": bson.M{"servers": server}}
-	_, err = db.staticDB.Collection(collSkylinks).UpdateOne(ctx, filter, update)
+	_, err := db.staticDB.Collection(collSkylinks).UpdateOne(ctx, filter, update)
+	return err
+}
+
+// FindAndLockUnderpinned fetches and locks a single underpinned skylink
+// from the database. The method selects only skylinks which are not pinned by
+// the given server.
+//
+// The MongoDB query is this:
+// db.getCollection('skylinks').find({
+//     "unpin": false,
+//     "$expr": { "$lt": [{ "$size": "$servers" }, 2 ]},
+//     "servers": { "$nin": [ "ro-tex.siasky.ivo.NOPE" ]},
+//     "$or": [
+//         { "lock_expires" : { "$exists": false }},
+//         { "lock_expires" : { "$lt": new Date() }}
+//     ]
+// })
+func (db *DB) FindAndLockUnderpinned(ctx context.Context, server string, minPinners int) (skymodules.Skylink, error) {
+	// First try to fetch a skylink which is locked by the current server. This
+	// is our mechanism for proactively recovering from files being left locked
+	// after a server crash.
+	sl, err := db.findLockedSkylink(ctx, server)
```

Comment:
We don't want to lock multiple skylinks because we want other servers to have a chance to pin them instead, thus spreading the additional load.

Also, the process you're proposing can potentially take a long time, depending on the combined size of the skylinks we're pinning.